File: PetHealth/app/src/main/java/pe/edu/upc/pethealth/activities/SignUpActivity.java
package pe.edu.upc.pethealth.activities;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.DatePickerDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.support.design.widget.TextInputEditText;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.Spinner;

import com.androidnetworking.AndroidNetworking;
import com.androidnetworking.common.Priority;
import com.androidnetworking.error.ANError;
import com.androidnetworking.interfaces.JSONObjectRequestListener;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.List;
import java.util.Date;
import pe.edu.upc.pethealth.R;
import pe.edu.upc.pethealth.models.DocumentType;
import pe.edu.upc.pethealth.network.PetHealthApiService;

public class SignUpActivity extends AppCompatActivity {

    Button doneButton;
    String tag;
    EditText usernameEditText;
    TextInputEditText passwordEditText;
    EditText emailEditText;
    final Context context = this;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sing_up);
        Toolbar myToolbar = (Toolbar) findViewById(R.id.my_toolbar);
        myToolbar.setNavigationIcon(R.drawable.ic_arrow_back_24dp);
        myToolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(view.getContext(), StartActivity.class);
                view.getContext().startActivity(intent);
            }
        });
        setSupportActionBar(myToolbar);
        usernameEditText =(EditText) findViewById(R.id.usernameEditText);
        passwordEditText = (TextInputEditText) findViewById(R.id.passwordTextInputEditText);

        emailEditText =(EditText) findViewById(R.id.emailEditText);

        doneButton = (Button) findViewById(R.id.doneButton);
        doneButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                attemptToSignUp();
            }
        });
        tag = "PetHealth";
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item){
        switch (item.getItemId()) {
            case android.R.id.home:
                this.finish();
                return  true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void attemptToSignUp() {
        //Reset Errors

        passwordEditText.setError(null);
        usernameEditText.setError(null);
        emailEditText.setError(null);


        //Store values

        String username = usernameEditText.getText().toString();
        String password = passwordEditText.getText().toString();
        String email = emailEditText.getText().toString();
        View focusView = null;
        Boolean cancel = false;

        if(TextUtils.isEmpty(username)) {
            usernameEditText.setError("This field is empty");
            focusView = usernameEditText;
            cancel = true;
        }
        if(TextUtils.isEmpty(password)) {
            passwordEditText.setError("This field is empty");
            focusView = passwordEditText;
            cancel = true;
        }
        if(TextUtils.isEmpty(email)) {
            emailEditText.setError("This field is empty");
            focusView = emailEditText;
            cancel = true;
        }

        if (cancel) {
            // There was an error; don't attempt login and focus the first
            // form field with an error.
            focusView.requestFocus();
        } else {
            JSONObject user = new JSONObject();
            try {
                user.put("username",username);
                user.put("password",password);
                user.put("mail",email);

            }catch (JSONException e){
                e.printStackTrace();
            }

            AndroidNetworking.post(PetHealthApiService.SIGNUP_USER)
                    .addJSONObjectBody(user)
                    .setTag(getString(R.string.app_name))
                    .setPriority(Priority.MEDIUM)
                    .build()
                    .getAsJSONObject(new JSONObjectRequestListener() {
                        @Override
                        public void onResponse(JSONObject response) {
                            // do anything with response
                            try {
                                if ("done".equalsIgnoreCase(response.getString("status"))) {
                                    AlertDialog.Builder builder = new AlertDialog.Builder(context);
                                    builder.setMessage("User Corretly Created");
                                    builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                                        @Override
                                        public void onClick(DialogInterface dialogInterface, int i) {
                                            dialogInterface.cancel();
                                            Intent intent = new Intent(context, StartActivity.class);
                                            context.startActivity(intent);
                                            finish();
                                        }
                                    });
                                    AlertDialog alertDialog = builder.create();
                                    alertDialog.show();
                                } else {
                                    Log.d(getString(R.string.app_name), "Error with the Resgistration of User");
                                }

                            } catch (JSONException e) {
                                e.printStackTrace();
                            }
                        }

                        @Override
                        public void onError(ANError anError) {
                            Log.d(getString(R.string.app_name), anError.getLocalizedMessage());
                        }
                    });
        }
    }
}

File: PetHealth/clinic-app/src/main/java/pe/edu/upc/phclinic/activities/SignUpActivity.kt
package pe.edu.upc.phclinic.activities

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.util.Log
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.github.salomonbrys.kotson.put
import com.google.android.gms.tasks.Continuation
import com.google.android.gms.tasks.Task
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.google.firebase.storage.UploadTask
import com.google.gson.JsonObject
import kotlinx.android.synthetic.main.activity_sing_up.*
import org.json.JSONException
import pe.edu.upc.phclinic.R
import pe.edu.upc.phclinic.network.RestClient
import pe.edu.upc.phclinic.network.RestView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class SignUpActivity : AppCompatActivity() {

    private lateinit var profileUri: Uri
    internal val context: Context = this
    lateinit var storageRef: StorageReference
    private val GALLERY_IMAGE = 1
    lateinit var username : String
    lateinit var password : String
    lateinit var mail : String
    lateinit var photo : String
    lateinit var userable_type : String
    lateinit var name : String
    lateinit var last_name : String
    lateinit var dni : String
    lateinit var phone : String



    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sing_up)
        storageRef = FirebaseStorage.getInstance().reference
        doneButton.setOnClickListener { attemptToSignUp() }
        profileImageView.setOnClickListener { chooseImage()}
    }
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when(requestCode){
            GALLERY_IMAGE -> {
                profileImageView.setImageURI(data?.data)
                profileUri = data!!.data
            }
        }
    }
    private fun chooseImage() {
        val intent = Intent(Intent.ACTION_GET_CONTENT)
        intent.type = "image/*"
        if (intent.resolveActivity(packageManager) != null) {
            startActivityForResult(intent, GALLERY_IMAGE)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                this.finish()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    private fun attemptToSignUp() {

        //Reset Errors
        passwordTextInputEditText.error = null
        usernameTextInputEditText.error = null
        emailTextInputEditText.error = null
        nameTextInputEditText.error = null
        lastNameTextInputEditText.error = null
        docNumberTextInputEditText.error = null
        phoneTextInputEditText.error = null


        var focusView: View? = null
        var cancel: Boolean? = false

        if (TextUtils.isEmpty(usernameTextInputEditText.text)) {
            usernameTextInputEditText.error = "This field is empty"
            focusView = usernameTextInputEditText
            cancel = true
        }
        if (TextUtils.isEmpty(passwordTextInputEditText.text)) {
            passwordTextInputEditText.error = "This field is empty"
            focusView = passwordTextInputEditText
            cancel = true
        }
        if (TextUtils.isEmpty(emailTextInputEditText.text)) {
            emailTextInputEditText.error = "This field is empty"
            focusView = emailTextInputEditText
            cancel = true
        }
        if (TextUtils.isEmpty(nameTextInputEditText.text)) {
            nameTextInputEditText.error = "This field is empty"
            focusView = nameTextInputEditText
            cancel = true
        }
        if (TextUtils.isEmpty(lastNameTextInputEditText.text)) {
            lastNameTextInputEditText.error = "This field is empty"
            focusView = lastNameTextInputEditText
            cancel = true
        }
        if (TextUtils.isEmpty(docNumberTextInputEditText.text)) {
            docNumberTextInputEditText.error = "This field is empty"
            focusView = docNumberTextInputEditText
            cancel = true
        }
        if (TextUtils.isEmpty(phoneTextInputEditText.text)) {
            phoneTextInputEditText.error = "This field is empty"
            focusView = phoneTextInputEditText
            cancel = true
        }

        if (cancel!!) {
            // There was an error; don't attempt login and focus the first
            // form field with an error.
            focusView!!.requestFocus()
        } else {
            prepareImage(usernameTextInputEditText.text){profileDownloadUri ->
                val bodyToSend = JsonObject()

                try {
                    bodyToSend.addProperty("username", username)
                    bodyToSend.addProperty("password", username)
                    bodyToSend.addProperty("mail", username)
                    bodyToSend.addProperty("photo", username)
                    bodyToSend.addProperty("userable_type", username)
                    bodyToSend.addProperty("name", username)
                    bodyToSend.addProperty("last_name", username)
                    bodyToSend.addProperty("dni", username)
                    bodyToSend.addProperty("phone", username)
                    bodyToSend.addProperty("linkedin_link", "")
                    bodyToSend.addProperty("degree", "")
                    bodyToSend.addProperty("location", "")
                    bodyToSend.addProperty("opening_hours", "")
                    bodyToSend.addProperty("website_url", "")
                    bodyToSend.addProperty("youtube_url", "")
                    bodyToSend.addProperty("twitter_url", "")
                } catch (e: JSONException) {
                    e.printStackTrace()
                }




            }
        }
    }
    private fun prepareImage(username: Editable?, callback: (profileDownloadUri: String)-> Unit) {
        val ref = storageRef.child("images/${username.toString()}")
        var uploadTask = ref.putFile(profileUri)

        uploadTask.continueWithTask(Continuation<UploadTask.TaskSnapshot, Task<Uri>> { task ->
            if (!task.isSuccessful) {
                task.exception?.let {
                    throw it
                }
            }
            return@Continuation ref.downloadUrl
        }).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val downloadUri = task.result
                callback(downloadUri.toString())
            }
        }
    }

    private fun sendUserData(body: JsonObject){
        val call = RestClient().service.signup(body)
        call.enqueue(object : Callback<RestView<JsonObject>>{
            override fun onResponse(call: Call<RestView<JsonObject>>, response: Response<RestView<JsonObject>>){
                val answer = response.body()

            }

            override fun onFailure(call: Call<RestView<JsonObject>>, t: Throwable) {

            }
        })



    }
}